src/functions/OverloadedOoops.kts

```diff
@@ -0,0 +1,29 @@
+package functions.overloadedOoops
+
+class Test {
+
+    fun test(block: () -> Int = { 0 }): String {
+        return "${block()}"
+    }
+
+    fun test2(block: () -> Int = { 0 }): () -> String {
+        return { test(block) }
+    }
+}
+
+
+{ println("Start") }()
+
+
+val t = Test()
+val a = t.test { 2 } == t.test2 { 2 }()
+val b = t.test() == t.test2()()
+
+
+{ println("Finish $a $b") }()
+
+// What will it print?
```

Comment:
Would you mind explaining the answer? Did you find this IRL?

Reply:
No, that code is not used in production. It's used to make puzzler looks like it's solve real case. I mean, if you write only 
`{ println("Finish $a $b") }`
without calling it, it's really strange a bit for what reason it looks so. 

So, that puzzler was found during development of android application. But construction like `{}()` is not from real life.

Reply:
Ah, ok, makes sense.
Btw, by "explaning the answer" I meant `Rationale.md`, it's quite a tricky one.